executor/executor.go

```diff
@@ -0,0 +1,602 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"sort"
+	"strings"
+
+	"fmt"
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/optimizer/evaluator"
+	"github.com/pingcap/tidb/optimizer/plan"
+	"github.com/pingcap/tidb/sessionctx/forupdate"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/types"
+)
+
+var (
+	_ Executor = &TableScanExec{}
+	_ Executor = &IndexScanExec{}
+	_ Executor = &IndexRangeExec{}
+	_ Executor = &SelectFieldsExec{}
+	_ Executor = &FilterExec{}
+	_ Executor = &LimitExec{}
+	_ Executor = &SortExec{}
+)
+
+// Row represents a record row.
+type Row struct {
+	// Data is the output record data for current Plan.
+	Data []interface{}
+
+	RowKeys []*RowKeyEntry
+}
+
+// RowKeyEntry is designed for Delete statement in multi-table mode,
+// we should know which table this row comes from.
+type RowKeyEntry struct {
+	// The table which this row come from.
+	Tbl table.Table
+	// Row handle.
+	Key string
+}
+
+// Executor executes a query.
+type Executor interface {
+	Fields() []*ast.ResultField
+	Next() (*Row, error)
+	Close() error
+}
+
+// TableScanExec represents a table scan executor.
+type TableScanExec struct {
+	t      table.Table
+	fields []*ast.ResultField
+	iter   kv.Iterator
+	ctx    context.Context
+}
+
+// Fields implements Executor Fields interface.
+func (e *TableScanExec) Fields() []*ast.ResultField {
+	return e.fields
+}
+
+// Next implements Execution Next interface.
+func (e *TableScanExec) Next() (row *Row, err error) {
+	if e.iter == nil {
+		var txn kv.Transaction
+		txn, err = e.ctx.GetTxn(false)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		e.iter, err = txn.Seek([]byte(e.t.FirstKey()))
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	if !e.iter.Valid() || !strings.HasPrefix(e.iter.Key(), e.t.KeyPrefix()) {
+		return
+	}
+	// TODO: check if lock valid
+	// the record layout in storage (key -> value):
+	// r1 -> lock-version
+	// r1_col1 -> r1 col1 value
+	// r1_col2 -> r1 col2 value
+	// r2 -> lock-version
+	// r2_col1 -> r2 col1 value
+	// r2_col2 -> r2 col2 value
+	// ...
+	rowKey := e.iter.Key()
+	handle, err := util.DecodeHandleFromRowKey(rowKey)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// TODO: we could just fetch mentioned columns' values
+	row = &Row{}
```

Comment:
If we should define return value in a function， i prefer to use `func (e *TableScanExec) Next() (*Row, error)` as definition.